distributed/protocol/serialize.py

```diff
@@ -132,8 +146,13 @@ def serialize(x, serializers=None, on_error="message", context=None):
     if isinstance(x, Serialized):
         return x.header, x.frames
 
+    # Check for "dask"-serializable data in dict/list/set
+    supported = (
+        isinstance(x, list) and "pickle" not in serializers
+    ) or check_dask_serializable(x)
```

Comment:
What's going on here?  I know that you added this because a particular test was failing, but what was the underlying cause?

Reply:
Sorry - Didn't have time to look into this carefully yet.  A list is being converted to a tuple during a round trip.  I originally thought it was a msgpack limitation (since the failing test only provides the "msgpack" serializer).  However, the problem may be something else in `client.publish_dataset` or `get_dataset` (a simple msgpack dumps-loads test works fine).

Reply:
The problem seems to be that the `use_list=False` argument is passed to `msgpack.loads` in [`msgpack_loads`](https://github.com/dask/distributed/blob/5243d23df9efb58d4f0d643c29a40062b365b65e/distributed/protocol/serialize.py#L72) - Not sure if there is a reason for this?

Reply:
Looks like that might be intentional. Appears to have been added in PR ( https://github.com/dask/distributed/pull/2000 ). @mrocklin, how would you like us to proceed here?

Reply:
Agreed @jakirkham - Removing that option allows tests to pass here, but may have uninteded consequences.

Reply:
Setting `use_list=False` will let it pass with `tuple`s, but fail with `list`s. Maybe we should be tracking the collection type in some other kind of metadata (if that is important to us)?

Reply:
> @mrocklin, how would you like us to proceed here?

Figure out what's going on (which it seems that you're doing), come up with an explanation of what's going on, make a recommendation, and then implement it?

I'm not sure if you're asking me to solve this problem for you or not.  If so, you might have to wait for a long while.  It looks like you all are making good progress though.

Reply:
> Setting use_list=False will let it pass with tuples, but fail with lists. Maybe we should be tracking the collection type in some other kind of metadata (if that is important to us)?

Maybe msgpack serialization just isn't capable of handling type/list type ambiguities?

(please don't take anything I say here too seriously, I haven't looked into this problem at all and don't really know what I'm talking about here)

Reply:
> Maybe msgpack serialization just isn't capable of handling type/list type ambiguities?

Yeah I think that is right.

> I'm not sure if you're asking me to solve this problem for you or not.

In some sense Rick already solved the problem, but it deviates from an earlier PR you submitted. So we are trying to understand if that deviation is ok.

Reply:
I think the best solution is to revert back to something similar to the original check that initiated this discussion, but to add a comment about the reasoning for the check (including a reference to PR#2000).

It is clear that msgpack will treat collections as either lists or tuples - Since we are assuming tuple support throughout, it is best to force "dask" serialization of lists in the case that "msgpack" is in `serializers`.

Reply:
It's true msgpack has issues.

```python
In [1]: import msgpack                                                          

In [2]: t = (0, 1, 2)                                                           

In [3]: l = [1, 2, 3]                                                           

In [4]: msgpack.loads(msgpack.dumps(t))                                         
Out[4]: [0, 1, 2]

In [5]: msgpack.loads(msgpack.dumps(l))                                         
Out[5]: [1, 2, 3]
```

Though Dask actually handles this correctly.

```python
In [1]: from distributed.protocol import serialize, deserialize                 

In [2]: t = (0, 1, 2)                                                           

In [3]: l = [1, 2, 3]                                                           

In [4]: deserialize(*serialize(t, serializers=["msgpack"]))                     
Out[4]: (0, 1, 2)

In [5]: deserialize(*serialize(l, serializers=["msgpack"]))                     
Out[5]: [1, 2, 3]
```

If one looks at the headers, it appears to be tracking the type information.

```python
In [6]: serialize(t, serializers=["msgpack"])                                   
Out[6]: 
({'sub-headers': [{'serializer': 'msgpack'},
   {'serializer': 'msgpack'},
   {'serializer': 'msgpack'}],
  'is-collection': True,
  'frame-lengths': [1, 1, 1],
  'type-serialized': 'tuple'},
 [b'\x00', b'\x01', b'\x02'])
```

Not sure why we are not seeing this behavior here. Is there still another issue lurking?

Reply:
Master is currently iterating through the list if it is <= 5 elements long (and directly labeling the type as `'type-serialized'`.  The goal here was to throw out the 5-element rule for something more rigorous.  So far, the rule is that the collection contains a dask-serializable object as the first element (nesting allowed), or it is a list and msgpack is an available serializer (which we know will produce a tuple).

Reply:
Good point! Wasn't paying attention to the size of the collections. Sounds like a bug in msgpack serialization/deserialization. Filed as issue ( https://github.com/dask/distributed/issues/3716 ).

Reply:
What does supported mean here?  It can be serialized?  If so, what if it is a list of unserializable objects?

Reply:
Supported just means it contains a data type registered by dask_serialize.  Sorry - Would you prefer a more descriptive name?

Reply:
But doesn't `[an_unsupported_object]` pass this check if the msgpack serializer is included?

Reply:
Yes - You are right. supported just means “we should iterate through the collection”

Reply:
@mrocklin - I just pushed some changes to (hopefully) make the logic a bit more clear.  For example, I am using the term `iterate_collection` instead of `supported`.   I am also now considering the order of the `serializers` list, because `"msgpack"` is only a problem for lists if it comes before `"pickle"` (I added `test_serialize_lists` to show/check this).

Reply:
Ah, I see what's going on now.  OK, thanks @rjzamora !

Reply:
With PR ( https://github.com/dask/distributed/pull/4575 ), I think we are able to workaround this MsgPack oddity. So dropped this code there.